README
INTRODUCTION
============

media-player-info is a repository of data files describing media player
(only USB Mass Storage ones) capabilities. These files contain information
about the directory layout to use to add music to these devices, about the
supported file formats, ... These capabilities used to be provided by HAL 
in the 10-usb-music-players.fdi file but had to be moved elsewhere as part
of the big HALectomy.

The music player capabilities are now described in .mpi files (which are 
.ini-like files) which you can find in media-players/. These mpi files are
used to generate an udev rule to identify these devices. This rule associate
an ID_MEDIA_PLAYER attribute to the media player devices. This attribute
specifies the name of the .mpi file to use to find out the device
capabilities. The .mpi can then be looked up in $XDG_DATA_DIRS/media-player-info
(see the XDG Base Directory Specification for more information about what
this means).

Both the mpi files and the udev rule have to be installed on the system.
Indeed, the udev rule only contain what is strictly necessary, ie it only
matches media player devices and set an ID_MEDIA_PLAYER udev property on 
the device. This property contains the name of the mpi file. It's up to 
applications needing detailed information about media players (like the audio 
formats they support, ...) to parse this file and extract whatever is necessary
from it. The media player information isn't included directly in udev database
because not many apps need it, and because we didn't want to repeat the same
mistakes as HAL and include everything in its database.

If you encounter an USB Mass Storage device that isn't detected correctly, 
please open a bug against media-player-info :
https://bugs.freedesktop.org/enter_bug.cgi?product=media-player-info


MPI FILE FORMAT
===============

Each mpi file is separated in several sections:
- [Device] (only mandatory section)
- [Media]
- [Playlist]
- [storage]


The [Device] section contains information about the media player:
- Product: human-readable product (device) name
- Vendor: human-readable vendor name
- AccessProtocol: the way this device is accessed. For now the only
  supported value is "storage" for USB mass storage devices.
- Icon: (optional) Icon name to use instead of the default "media-player".

For actually identifying a player from hardware information in /sys, there are
two possibilities:

- DeviceMatch field: semi-colon separated list of "usb:<vid>:<pid>" values
  (<vid> is the USB vendor ID, <pid> is the USB product ID). This lists all the
  USB that are described by the current mpi file.

- A combination of "USBVendor", "USBProduct", "USBModel", and "USBManufacturer"
  strings (which map to the corresponding sysfs fields). You can use the '*'
  wildcard. See apple-ipod.mpi for an example.

The [Media] section describes in more details which media formats the player
supports. For now it only describes audio capabilities, but video or artwork
capabilities might be added in the future. Each entry in the Media section
is a semi-colon-separated list of media mime types (audio/mpeg, audio/x-wav,
...).
- InputFormats: audio formats the media player can record to
- OutputFormats: audio formats the media player can play back


The [Playlist] section describes the player playlist capabilites. If it's
absent, it means the player doesn't support playlists (or that the information
contained in media-player-info is incomplete, in such a case, please email us
or file a bug).
- Formats: a semi-colon-separated list of the mime-types of the
  playlist formats supported by the device
- FolderSeparator: separator between directory/file names in playlist files. If
  not given, "Unix" is assumed, which uses '/'. Some players require 
  "DOS", which will use '\'. Other values are invalid and will be ignored.
- LineEnding: line separator in playlist files. This can have the values 'CR',
  'LF', or 'CRLF', which correspond to 0x0D (\r), 0x0A (\n), and 0x0D0A (\r\n)
  respectively. If not given, the default behaviour depends on the music player
  software.

The [storage] section describes the location where media files can be found and
have to be stored.
- AudioFolders: a semi-colon-separeted list of directories where audio files
  can be found and must be stored
- PlaylistPath: describes where and how the playlist files can be found and
  have to be stored 
- RequiresEject: the eject ioctl is required to properly eject the media
- FolderDepth: this tells applications exactly how deep of directory
  hierarchies files should be placed in. If the device does not have a
  limit, do not set this property.


MPI FILES NAMING CONVENTION
===========================
The current naming convention for .mpi file is as follows :

<vendor>_<device>.mpi

Where :
- <vendor> is the official name of the device's manufacturer. If the same
  usb id is used by different market names, use the manufacturer name from
  usb.ids/lsusb. If vendor name has more than one word, words are separated by
  '-'
- <device> is the official name of the device. If the same .mpi file matches
  more than one device, you can separate devices with extra "_". If the .mpi
  file matches a lot of different market names, you can use 0x1111 where 1111
  is the product id of the device (in this case, use a generic name inside the
  .mpi file (e.g. "Mobile phone", "Media player"). If device name has more than
  one word, words are separated by '-'

Please note that these conventions are not absolute rules but we try to keep
the naming coherant.



NEWS
media-player-info 22 (2014-10-07)
=================================
Martin Pitt:
      Terminate udev rules with an empty line (LP #1364747)
      Add Sandisk Sansa Clip Sport (LP #1296418)
      Remove ZTE Score (Debian #748647)

media-player-info 21 (2013-07-24)
=================================
Martin Pitt (6):
      Fix build under non-UTF8 locales (#67242)
      Fix build in separate build tree

media-player-info 19 (2013-07-23)
=================================
Martin Pitt (5):
      Drop obsolete fdi2mpi.py
      Port generator tools to Python 3
      Rename mpi files with overly long names
      Add Motorola Droid 4

Tom Gundersen (1):
      Create udev hwdb when building with udev >= 196

media-player-info 18 (2013-07-05)
=================================
Dan Williams (1):
      Add HP Veer

Markus Ottensmann (1):
      Add Archos 5 Internet Tablet

Martin Pitt (3):
      Add ZTE Score
      Add LG Optimus One P500
      Drop newer kindles

Pavol Babincak (1):
      Fix playlist path for Sony Ericsson Xperia mini pro

media-player-info 17 (2012-05-30)
=================================
Jason Leonard (1):
      Add Creative Zen MX

Jonathan Ernst (6):
      Xperia X12 is better known as Xperia arc
      Add Nec LifeTouch Note
      Galaxy Gio (GT-S5660) shares same is as Ace
      Add a third ID for Acer Liquid (MT)
      Add Nokia C7-00
      Add Sony Ericsson Xperia Mini

Linus Walleij (1):
      Add support for iBeat Jess

Martin Pitt (3):
      Add Nokia N9
      Add playlist format to Android devices
      Add alternative product ID for Sony Ericsson Xperia X10

media-player-info 16 (2012-01-24)
=================================
Alex Fiestas (1):
      Add Nokia N950

Martin Pitt (8):
      Fix Nokia N950
      Add Sandisk Sansa Clip ZIP
      Add Sony Walkman NWZ-S765
      Generalize Sony NWZ-S76x
      Add Sony Walkman S-750 series
      Add Sony Walkman NWZ-S545
      Add Archos Key
      Add Sony Walkman NWZ-A800 series

media-player-info 15 (2011-10-04)
=================================
Alejandro Mery (1):
      Add another product ID for Sansa Clip+

Bernd Schlapsi (1):
      Add audio folder for Sony Walkman NWZ-S638F

Jo Shields (2):
      Add HP Pre 3
      Add HP Touchpad

Jonathan Ernst (21):
      Add Sony Ericsson W705
      Add Vodafone 845 aka Huawei U8120 Android phone
      Add LG Optimus S Android phone
      Add Sharp IS03 Android phone
      Add Sony Ericsson Xperia X12 (Arc) Android phone
      Add Samsung Galaxy S2
      Add Samsung Reality
      0x05c6_0x1000 is also used by Samsung Freeform II not only LG phones
      Add another ID for Samsung Galaxy S2 (thanks kazken3) for the hint
      Add GoGear Vibe 4Gb (Modelo SA2VBE)
      Add Nokia 6730 Classic
      Add Philips GoGear Raga MP3 player
      Add Sony Ericsson Vivaz
      Add Barnes and Noble Nook eReaders
      Add Motorola Atrix
      Add Archos A43 Internet Tablet
      Add Notion Ink Adam Android Tablet
      Add Samsung Galaxy Ace
      Add Acer Liquid
      Add Cowon iAUDIO E2
      Fix typo

Martin Pitt (10):
      Add TrekStor i.Beat rock
      Add playlist path to all Android devices
      Remove %File references
      SanDisk Sansa Clip+: Add alternate podcasts directory
      Add Samsung Galaxy SII
      Add Palm Pre 2
      Define "FolderSeparator" and "LineEnding" properties
      Sandisk Sansa Fuze: Set FolderSeparator
      Use symbolic names for FolderSeparator
      Add Creative Zen Style 300

Pavol Babinčák (1):
      Add Sony Ericsson Xperia mini pro

Shishir Joshi (1):
      Add MicroMax A60 (Android based)

William Witt (2):
      Add Galaxy S 2
      sandisk_sansa-fuze-plus: Add MSC mode Device ID

Wojciech Zylinski (1):
      Add Sony Ericsson Xperia Play

media-player-info 14 (2011-04-19)
=================================
Alejandro Mery (1):
      Add Sandisk Sansa Fuze+

Jonathan Ernst (4):
      HTC legend uses same ID as desire/hero/evo-4g in debug mode
      Add Pantech SIRIUS alpha
      Add Sharp IS01
      Add Sony Ericsson C905

Martin Pitt (2):
      Fix product ID of Sandisk Sansa Fuze.
      Add Creative Zen X-Fi Style

media-player-info 13 (2011-03-15)
=================================
Jonathan Ernst (18):
      HTC Evo uses same ID as Desire/Hero in non debug mode -> merging
      Vortex uses same ID (in debug mode) as Ally -> merge
      Add playlist format to HTC Desire and friends
      Improve LG 0x6018 phones m-p-i file
      Improve Sony Ericsson Xperia X10 mini Pro m-p-i
      Add Sony Ericsson XPERIA X1
      Add Motorola Droid 2 Android phone
      Rename Motorola Droid X mpi file
      Fix USB id of sony-ericsson K800i (was phone mode instead of mass storage mode)
      Add storage folder to Sony Ericsson K800i
      Add Huawei Ideos non-debug mode
      Add Samsung Galaxy S debug ID
      Add Desire HD debug ID
      Don't generate icon rules as it is currently broken (see LP #657609)
      Add 5 Coby players
      Add FolderDepth=2 for the Zen Stone
      Add RCA H100a
      Add informations about bug reporting

Martin Pitt (1):
      Add Samsung U6.

Philippe Gauthier (1):
      Add Sony Walkman NWZ-E355.

media-player-info 12 (2011-01-08)
=================================
Jonathan Ernst (19):
      Improve Blackberries definition
      Add Playlist formats supported by the Samsung Galaxy
      Add ID for Blackberry 8130 Pearl
      Add new blackberries     Merge all blackberries as the ID depends not on t
      Add Samsung YP-P2
      Add newer Kindles
      Add Nokia 6700
      Add Samsung GT-S8000 Jet
      Add LG Ally
      Use friendly vendor name for LG
      Add non debug USB ID for Ally and improve mpi
      Fix typo in Philips GoGear Opus mpi
      Add Philips GoGear Aria
      Merge LG Ally and Optimus One as they use the same usb id.
      Use naming convention for apple-video-ipod.mpi file
      Add LG Vortex
      Add HTC EVO 4G
      Add Google Nexus S
      Add Google Nexus S's debug mode usb id

Luca Ferretti (1):

tools/mpi2hwdb.py
#!/usr/bin/env python3
# Generate hwdb file from music player identification (.mpi) files
#
# (C) 2009 Canonical Ltd.
# (C) 2013 Tom Gundersen <teg@jklm.no>
# Author: Tom Gundersen <teg@jklm.no>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.

import sys, configparser, os.path, os

def parse_mpi(mpi):
    '''Print hwdb file for given ConfigParser object.'''

    cp = configparser.RawConfigParser()
    assert cp.read(mpi, encoding='UTF-8')

    # if we have more info than just idVendor+idProduct we need to use an udev rule,
    # so don't write an hwdb entry
    for name in ['usbvendor', 'usbproduct', 'usbmodel', 'usbmanufacturer']:
        try:
            cp.get('Device', name)
            return
        except configparser.NoOptionError:
            continue

    block = ''

    try:
        m = cp.get('Device', 'product')
        block += '# ' + m + '\n'
    except configparser.NoOptionError:
        pass

    try:
        usbids = {}
        for usbid in cp.get('Device', 'devicematch').split(';'):
            if len(usbid.split(':')) != 3:
                continue
            (subsystem, vid, pid) = usbid.split(':')
            if subsystem != "usb":
                continue
            if vid in usbids:
                usbids[vid].append(pid)
            else:
                usbids[vid] = [ pid ]

        for vid, pids in usbids.items():
            for pid in pids:
                block += 'usb:v%sp%s*\n' % (vid.upper(), pid.upper())
                block += ' ID_MEDIA_PLAYER=%s\n' % os.path.splitext(os.path.basename(mpi))[0]

                # do we have an icon?
                try:
                    icon = cp.get('Device', 'icon')
                    # breaks media player detection : https://bugs.launchpad.net/ubuntu/+source/gvfs/+bug/657609
                    #print ' UDISKS_PRESENTATION_ICON_NAME=%s\n' % icon,
                except configparser.NoOptionError:
                    pass

                # empty line between blocks
                block += '\n'

                # write bytes so that we are independent of the locale
                os.write(sys.stdout.fileno(), block.encode('UTF-8'))

    except configparser.NoOptionError:
        pass

#
# main
#

# parse MPI files
for f in sys.argv[1:]:
    parse_mpi(f)


tools/mpi2udev.py
#!/usr/bin/env python3
# Generate udev rules from music player identification (.mpi) files
#
# (C) 2009 Canonical Ltd.
# Author: Martin Pitt <martin.pitt@ubuntu.com>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.

import sys, configparser, os.path, os

# translation of mpi Device keys to udev attributes
mpi2udev = {
    'vendorid': 'ATTRS{idVendor}=="%s"',
    'productid': 'ATTRS{idProduct}=="%s"',
    'usbvendor': 'ATTRS{vendor}=="%s"',
    'usbmodel': 'ATTRS{model}=="%s"',
    'usbproduct': 'ATTRS{product}=="%s"',
    'usbmanufacturer': 'ATTRS{manufacturer}=="%s"',
}

def parse_mpi(mpi, hwdb):
    '''Print udev rule for given ConfigParser object.'''

    cp = configparser.RawConfigParser()
    assert cp.read(mpi, encoding='UTF-8')

    rule = ''

    for name in ['usbvendor', 'usbproduct', 'usbmodel', 'usbmanufacturer']:
        try:
            value = cp.get('Device', name)
            rule += mpi2udev[name] % value + ', '
        except configparser.NoOptionError:
            continue


    # if using hwdb and no info other than idVendor+idProduct was found, we don't need to write
    # an udev rule
    if hwdb and not rule:
        return

    try:
        usbids = {}
        for usbid in cp.get('Device', 'devicematch').split(';'):
            if len(usbid.split(':')) != 3:
                continue
            (subsystem, vid, pid) = usbid.split(':')
            if subsystem != "usb":
                continue
            if vid in usbids:
                usbids[vid].append(pid)
            else:
                usbids[vid] = [ pid ]

        for vid, pids in usbids.items():
            rule += 'ATTRS{idVendor}=="%s", ATTRS{idProduct}=="%s"'% (vid, '|'.join(pids)) + ', '
    except configparser.NoOptionError:
        pass

    # if no information was found, don't write a rule at all
    if not rule:
        return

    try:
        m = cp.get('Device', 'product')
        rule = ('# %s\n' % m) + rule
    except configparser.NoOptionError:
        pass

    rule += 'ENV{ID_MEDIA_PLAYER}="%s"' % os.path.splitext(os.path.basename(mpi))[0]

    # do we have an icon?
    try:
        icon = cp.get('Device', 'icon')
        # breaks media player detection : https://bugs.launchpad.net/ubuntu/+source/gvfs/+bug/657609
        # rule += ', ENV{UDISKS_PRESENTATION_ICON_NAME}="%s"' % icon
    except configparser.NoOptionError:
        pass

    # empty line between blocks
    rule += '\n\n'

    # write bytes so that we are independent of the locale
    os.write(sys.stdout.fileno(), rule.encode('UTF-8'))

#
# main
#

# udev rules header
os.write(sys.stdout.fileno(), b'''ACTION!="add|change", GOTO="media_player_end"
# catch MTP devices
ENV{DEVTYPE}=="usb_device", GOTO="media_player_start"

# catch UMS devices
SUBSYSTEM!="block", GOTO="media_player_end"
SUBSYSTEMS=="usb", GOTO="media_player_start"
GOTO="media_player_end"

LABEL="media_player_start"

''')

# the first argument should be "hwdb" or "udev"
hwdb = False
if sys.argv[1] == 'hwdb':
    hwdb = True
elif sys.argv[1] != 'udev':
    sys.stderr.write('Unknown mode %s. Must be "hwdb" or "udev".\n' % sys.argv[1])
    sys.exit(1)

# parse MPI files
for f in sys.argv[2:]:
    parse_mpi(f, hwdb)

# udev rules footer
os.write(sys.stdout.fileno(), b'\nLABEL="media_player_end"\n')


tools/udev-syntax-check.py
#!/usr/bin/env python3
# Simple udev rules syntax checker
#
# (C) 2010 Canonical Ltd.
# Author: Martin Pitt <martin.pitt@ubuntu.com>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with keymap; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.

import re
import sys

if len(sys.argv) < 2:
    sys.stderr.write('Usage: %s <rules file> [...]\n' % sys.argv[0])
    sys.exit(2)

no_args_tests = re.compile('(ACTION|DEVPATH|KERNELS?|NAME|SYMLINK|SUBSYSTEMS?|DRIVERS?|TAG|PROGRAM|RESULT)\s*(?:=|!)=\s*"([^"]*)"$')
args_tests = re.compile('(ATTRS?|ENV|TEST){(\w+)}\s*(?:=|!)=\s*"([^"]*)"$')
no_args_assign = re.compile('(NAME|SYMLINK|OWNER|GROUP|MODE|TAG|RUN|LABEL|GOTO|WAIT_FOR|OPTIONS)\s*(?:\+=|:=|=)\s*"([^"]*)"$')
args_assign = re.compile('(ATTR|ENV|IMPORT){([a-zA-Z0-9_.-]+)}\s*=\s*"([^"]*)"$')

result = 0
lineno = 0
for path in sys.argv[1:]:
    for line in open(path, encoding='UTF-8'):
        lineno += 1

        # filter out comments and empty lines
        line = line.split('#')[0]
        line = line.strip()
        if not line:
            continue

        for clause in line.split(','):
            clause = clause.strip()
            if not (no_args_tests.match(clause) or args_tests.match(clause) or
                    no_args_assign.match(clause) or args_assign.match(clause)):

                sys.stderr.write('Invalid line %s:%i: "%s"\n' % (path, lineno, line))
                sys.exit(1)
                result = 1
                break

sys.exit(result)


tools/fdi2mpi.py
#!/usr/bin/python
# Convert hal's USB music player FDIs into media player information (.mpi)
# property files.
#
# (C) 2009 Canonical Ltd.
# Author: Martin Pitt <martin.pitt@ubuntu.com>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
#
#
# Alternatively this script can be distributed under the X11 license:
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is fur-
# nished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,  FIT-
#  NESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
#  AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,  WHETHER
#  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN  CON-
#  NECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


import sys, os.path, xml.dom.minidom

# translation of hal match keys to mpi attribute names
_hal_match2mpi = {
    '@storage.originating_device:usb.vendor_id': 'VendorID',
    '@storage.physical_device:usb.vendor_id': 'VendorID',
    '@storage.originating_device:usb.product_id': 'ProductID',
    '@storage.physical_device:usb.product_id': 'ProductID',
    '@storage.originating_device:@info.parent:usb_device.product': 'USBModel',
    'storage.vendor': 'USBVendor',
    'storage.model': 'USBModel',

    # if we actually want/need to check for subdevices, we can do:
    #'storage.lun': 'KERNELS=="*:*:*:%s"', # host:bus:target:lun
    # ... but let's ignore it for now:
    'storage.lun': None,

    # the following are globally checked by rule header, ignore them 
    '@storage.originating_device:info.subsystem': None,
    'info.category': None,
}

_hal_prop2mpi = {
    'portable_audio_player.access_method.protocols': ('Device', 'AccessProtocol'),
    'portable_audio_player.access_method': ('Device', 'AccessProtocol'),
    'portable_audio_player.output_formats': ('Media', 'OutputFormats'),
    'portable_audio_player.input_formats': ('Media', 'InputFormats'),
    'portable_audio_player.audio_folders': ('storage', 'AudioFolders'),
    'portable_audio_player.folder_depth': ('storage', 'FolderDepth'),
    'portable_audio_player.playlist_path': ('storage', 'PlaylistPath'),
    'portable_audio_player.playlist_format': ('Playlist', 'Formats'),
    'portable_audio_player.playlist_formats': ('Playlist', 'Formats'),
    'storage.requires_eject': ('storage', 'RequiresEject'),
    'storage.model': ('Device', 'Model'),
    'storage.vendor': ('Device', 'Vendor'),
    'storage.drive_type': ('storage', 'DriveType'),

    # all of those are "portable audio player"
    'info.capabilities': (None, None),

    # not useful; if we need the device, we should set up a general rule to
    # slam it into the udev db
    'portable_audio_player.storage_device': (None, None),
}

def match_op2glob(node):
    '''Convert FDI match operator to a glob.
    
    Return glob
    '''
    if node.attributes.has_key('string'):
        return node.attributes['string'].nodeValue

    if node.attributes.has_key('contains'):
        return '*' + node.attributes['contains'].nodeValue + '*'

    if node.attributes.has_key('int'):
        v = node.attributes['int'].nodeValue
        if v != '1':
            v = '%04x' % int(v, 0)
	return v

    if node.attributes.has_key('int_outof'):
        alternatives = node.attributes['int_outof'].nodeValue.split(';')
        hex_alternatives = []
        for a in alternatives:
            hex_alternatives.append('%04x' % int(a, 0))
	return hex_alternatives

    if node.attributes.has_key('contains_ncase'):
        v = node.attributes['contains_ncase'].nodeValue 
        nocase_glob = ''.join(['[%s%s]' % (c.lower(), c.upper()) for c in v])
	return '*' + nocase_glob + '*'

    if node.attributes.has_key('prefix_ncase'):
        v = node.attributes['prefix_ncase'].nodeValue 
        nocase_glob = ''.join(['[%s%s]' % (c.lower(), c.upper()) for c in v])
	return nocase_glob + '*'

    raise NotImplementedError, 'unknown string operator ' + str(node.attributes.keys())

def get_node_comment(node):
    '''Return comment before a node.'''

    while True:
        node = node.previousSibling
        if not node:
            break
        if node.nodeType == xml.dom.Node.COMMENT_NODE:
            lines = [l.strip() for l in node.nodeValue.splitlines()]
            return ' '.join(lines)
        if node.nodeType != xml.dom.Node.TEXT_NODE:
            break
    return None

def collect_attributes(node, attrs):
    '''Add media player attributes to attrs.'''

    for n in node.childNodes:
        if n.nodeType != xml.dom.minidom.Node.ELEMENT_NODE:
            continue
        if n.tagName == 'match':
            continue
        assert n.tagName in ('addset', 'merge', 'append')

        (cat, prop) = _hal_prop2mpi[n.attributes['key'].nodeValue]
        if cat is None:
            continue
        content_node = n.childNodes[0]
        assert content_node.nodeType == xml.dom.Node.TEXT_NODE

        attrs.setdefault(cat, {}).setdefault(prop, []).append(content_node.nodeValue.strip())

def mkfilename(attrs, current_vendor):
    '''Return an appropriate mpi file name for attributes.'''

    v = None
    if 'VendorID' in attrs.get('Device', {}):
        v = attrs['Device']['VendorID']
    if 'USBVendor' in attrs.get('Device', {}):
        v = attrs['Device']['USBVendor']
    if current_vendor:
        v = current_vendor.split()[0].split('/')[0].replace(',', '').lower().strip()
    assert v

    m = None
    if 'ProductID' in attrs.get('Device', {}):
        m = attrs['Device']['ProductID']
    if 'USBModel' in attrs.get('Device', {}):
        m = attrs['Device']['USBModel']
    # if we only have a single product ID, attempt to get nicer name
    if len(attrs['Device'].get('ProductID', [''])) <= 1 and \
            'Product' in attrs.get('Device', {}) and \
            '/' not in attrs['Device']['Product']:
        m = attrs['Device']['Product']
    if 'Model' in attrs.get('Device', {}):
        m = attrs['Device']['Model'][0]

    if type(m) == type([]):
        m = '_'.join(m)
    m = m.replace(' ', '_').replace(';', '_').replace('*', '').replace('(',
            '').replace(')', '').lower()
    assert m

    assert '*' not in m, m
    assert '[' not in m, m
    assert ';' not in m, m
    assert '(' not in m, m
    assert ')' not in m, m
    assert ' ' not in m, m

    # a lot of product names start with vendor name
    if m.startswith(v):
        m = m[len(v):]
    if m.startswith('_'):
        m = m[1:]

    print attrs['Device'], '->', '"%s-%s"' % (v, m)

    return '%s-%s' % (v, m)

def usb_ids_to_device_match(attrs):
    '''convert the USB IDs we gathered to a DeviceMatch string appropriate
    for mpi files'''

    if not 'VendorID' in attrs['Device'].keys():
	return

    if not 'ProductID' in attrs['Device'].keys():
	attrs['Device']['ProductID'] = [ '*' ]

    device_match = ''
    for id in attrs['Device']['ProductID']:
	device_match += 'usb:'+ attrs['Device']['VendorID'][0] + ':' + id + ';'

    attrs['Device']['DeviceMatch'] = device_match
    del attrs['Device']['VendorID']
    del attrs['Device']['ProductID']


def write_mpi(attrs, filename):
    '''Write mpi file for attrs.'''

    # define order of sections
    sections = ['Device', 'Media', 'Playlist', 'storage']

    # a lot of product names start with vendor name
    if attrs['Device'].has_key('Vendor') and attrs['Device'].has_key('Product'):
	product = attrs['Device']['Product']
	vendor = attrs['Device']['Vendor']
	if product.startswith(vendor):
	    attrs['Device']['Product'] = product[len(vendor):].strip()

    usb_ids_to_device_match(attrs)

    # only keep the most specific model name
    if attrs['Device'].has_key('Model'):
        attrs['Device']['Model'] = attrs['Device']['Model'][0]

    assert set(attrs.keys()) <= set(sections)

    f = open(os.path.join('media-players', filename + '.mpi'), 'w')

    for section in sections:
        if section not in attrs:
            continue
        print >> f, '[%s]' % section
        for k, v in attrs[section].iteritems():
            if type(v) == type([]):
                v = ';'.join(v)
            print >> f, '%s=%s' % (k, v)
        print >> f
        
    f.close()

def parse_leaf_match(node, current_vendor, current_model):
    '''Parse leaf match.

    Construct udev matches from all parent matches and then write property mpi
    file from all attached properties.
    '''
    # this is from the "Set common keys for detected audio player" section,
    # which we want to ignore
    if node.attributes['key'].nodeValue == 'portable_audio_player.access_method.protocols':
        return

    attrs = {'Media': {'OutputFormats': ['audio/mpeg']}}

    n = node
    while n:
        collect_attributes(n, attrs)

        key = n.attributes['key'].nodeValue
        device_attr = _hal_match2mpi[key]
        if device_attr:
            glb = match_op2glob(n)
	    if type(glb) == type([]):
                attrs.setdefault('Device', {}).setdefault(device_attr, []).extend(glb)
	    else:
                attrs.setdefault('Device', {}).setdefault(device_attr, []).append(glb)
        while True:
            n = n.parentNode
            if n is None or (n.nodeType == xml.dom.minidom.Node.ELEMENT_NODE and 
                    n.tagName == 'match'):
                break

    if current_model:
        attrs.setdefault('Device', {})['Product'] = current_model
    if current_vendor:
        attrs.setdefault('Device', {})['Vendor'] = current_vendor

    fname = mkfilename(attrs, current_vendor)
    write_mpi(attrs, fname)

def parse_fdi(fdi):
    '''Parse music player fdi node.'''

    current_vendor = None

    # find all <match> leaf nodes
    for match_node in fdi.getElementsByTagName('match'):
        current_model = None

        c = get_node_comment(match_node)

        # filter out a few weird cases in current hal-info
        if c and ('No-name' in c or 'require' in c or 'media files' in c):
            c = None
	if c and ('TODO' in c):
            c = 'Apple'
        if c and ('MegaScreen' in c):
            c = 'Nexia'
        if c and ('PSP' in c):
            c = 'Sony'

        if c:
            match_key = match_node.attributes['key'].nodeValue
            if 'model' in match_key or 'product' in match_key:
                current_model = c
            elif 'vendor' in match_key:
                current_vendor = c
            else:
                if 'subsystem' in match_key or 'protocols' in match_key or 'category' in match_key:
                    pass
                else:
                    raise NotImplementedError, 'do not know how to interpret comment ' + c

        children = set([n.tagName for n in match_node.childNodes 
                if n.nodeType == xml.dom.minidom.Node.ELEMENT_NODE])

        if u'match' in children:
            continue

        parse_leaf_match(match_node, current_vendor, current_model)

#
# main
#

for f in sys.argv[1:]:
    parse_fdi(xml.dom.minidom.parse(f))
